### DISCUSSION TITLE
Window CloseRequested issue

### DISCUSSION THREAD

**Discussion by Blatman** [tauri-apps/tauri#5334] [Q&A] [answered]
Having trouble controlling the window close event.
As I understand it the Rust BE code is able to detect if JS is listening for the window CloseRequested event.

My JS indeed does detect the `tauri://close-requested` event at the FE and will open a confirm dialog however the window will just close anyway.

I notice that the event payload is null which I would have expected to contain maybe label/signal data from the BE and the BE will pass control of the closing to the JS FE.

I'm not sure if the BE is detecting the listener at all - the window object shows the event but not as part of the .listeners node.

How does the Rust code detect the listening action? Does it need something to link it or does the rust look at the window object for listeners?

Looking at the rust src in manager.rs line 85 looks like emit and trigger will occur after the signal_tx is set true - not sure if this line will close the window anyway or pass control to the JS FE.

JS 
```js
var mainWinEvent = window.__TAURI__.event;
	
....

mainWinEvent.listen('tauri://close-requested', async function (event) {	
// event.payload is the payload object
console.log('EEE', event);
    const confirmed = await dialog.confirm('Are you sure?');
    if (!confirmed) {
        // user did not confirm closing the window; let's prevent it
//         event.preventDefault();   // `is not a function` error
	    // close main window after 4 sec
	    setTimeout(function() {
	        window.close();	
	    }, 4000);
console.log('Q result - false', confirmed);
    } else {
	    window.close();	
console.log('Q result - true', confirmed);
    }
})
```

JS event listen result-
<img width="1038" alt="prefsWin - CR" src="https://user-images.githubusercontent.com/11089335/193548549-0bec462d-3449-4e7d-89b0-5865a5ef35fa.png">

Window object -
<img width="641" alt="windowObject" src="https://user-images.githubusercontent.com/11089335/193548616-97727a31-db97-4d08-9086-6b28d26144d1.png">

Rust code -
<img width="870" alt="Tauri - managerDotrs - CR" src="https://user-images.githubusercontent.com/11089335/193548644-ce2c5c5a-9673-479e-ba22-aa0660141eb3.png">

**Comment by FabianLars** [comment, marked as answer]
Listening to the close-requested event only works on the actual WebviewWindow instance, and not with the global listen function. Would look something like this instead
```ts
window.__TAURI__.window.getCurrent().listen("tauri://close-requested", () => {});
```
We do have a helper function for this too: https://tauri.app/v1/api/js/window#oncloserequested

**Reply by Blatman** [reply]
Ah yes - the trick - use .getCurrent()

This code worked for me - all subwindows closed but the main was controlled by the dialog result.

JS -

    const appWindow = window.__TAURI__.window;
appWindow.getCurrent().listen("tauri://close-requested", async function (event) {	
console.log('EEE', event);
        const confirmed = await dialog.confirm('Are you sure?');
        if (!confirmed) {
	        // CANCEL button clicked - close main window after 4 sec of doing stuff
	        setTimeout(function() {
console.log('appWindow', appWindow);
console.log('Current appWindow', appWindow.getCurrent());
                    appWindow.getCurrent().close();
	        }, 4000);
console.log('Q result - false', confirmed);
        } else {
	    // OK button clicked - close now
            appWindow.getCurrent().close();
			console.log('Q result - true', confirmed);
        }
	})

Note:  appWindow.close() did not work but appWindow.getCurrent().close() does.
